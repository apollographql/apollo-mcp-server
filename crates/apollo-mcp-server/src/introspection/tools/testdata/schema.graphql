scalar DateTime
scalar JSON
scalar Upload

enum UserRole {
  ADMIN
  MODERATOR
  USER
  GUEST
}

enum ContentStatus {
  DRAFT
  PUBLISHED
  ARCHIVED
  DELETED
}

enum NotificationPriority {
  LOW
  MEDIUM
  HIGH
  URGENT
}

enum MediaType {
  IMAGE
  VIDEO
  AUDIO
  DOCUMENT
}

interface Node {
  id: ID!
  createdAt: DateTime!
  updatedAt: DateTime!
}

interface Content {
  id: ID!
  title: String!
  status: ContentStatus!
  author: User!
  metadata: JSON
}

type User implements Node {
  id: ID!
  createdAt: DateTime!
  updatedAt: DateTime!
  username: String!
  email: String!
  role: UserRole!
  profile: UserProfile
  posts: [Post!]!
  comments: [Comment!]!
  notifications: [Notification!]!
  preferences: UserPreferences!
}

type UserProfile {
  firstName: String
  lastName: String
  bio: String
  avatar: Media
  socialLinks: [SocialLink!]!
  location: Location
}

type Location {
  country: String!
  city: String
  coordinates: Coordinates
}

type Coordinates {
  latitude: Float!
  longitude: Float!
}

type SocialLink {
  platform: String!
  url: String!
  verified: Boolean!
}

type Post implements Node & Content {
  id: ID!
  createdAt: DateTime!
  updatedAt: DateTime!
  title: String!
  content: String!
  status: ContentStatus!
  author: User!
  metadata: JSON
  comments: [Comment!]!
  media: [Media!]!
  tags: [Tag!]!
  analytics: PostAnalytics!
}

type Comment implements Node {
  id: ID!
  createdAt: DateTime!
  updatedAt: DateTime!
  content: String!
  author: User!
  post: Post!
  parentComment: Comment
  replies: [Comment!]!
  reactions: [Reaction!]!
}

type Media {
  id: ID!
  type: MediaType!
  url: String!
  thumbnail: String
  metadata: MediaMetadata!
  uploader: User!
}

type MediaMetadata {
  size: Int!
  format: String!
  dimensions: Dimensions
  duration: Int
}

type Dimensions {
  width: Int!
  height: Int!
}

type Tag {
  id: ID!
  name: String!
  slug: String!
  description: String
  posts: [Post!]!
}

type Reaction {
  id: ID!
  type: String!
  user: User!
  comment: Comment!
  createdAt: DateTime!
}

type Notification {
  id: ID!
  type: String!
  priority: NotificationPriority!
  message: String!
  recipient: User!
  read: Boolean!
  createdAt: DateTime!
  metadata: JSON
}

type PostAnalytics {
  views: Int!
  likes: Int!
  shares: Int!
  comments: Int!
  engagement: Float!
  demographics: Demographics!
}

type Demographics {
  ageGroups: [AgeGroup!]!
  locations: [LocationStats!]!
  devices: [DeviceStats!]!
}

type AgeGroup {
  range: String!
  percentage: Float!
}

type LocationStats {
  country: String!
  count: Int!
}

type DeviceStats {
  type: String!
  count: Int!
}

type UserPreferences {
  theme: String!
  language: String!
  notifications: NotificationPreferences!
  privacy: PrivacySettings!
}

type NotificationPreferences {
  email: Boolean!
  push: Boolean!
  sms: Boolean!
  frequency: String!
}

type PrivacySettings {
  profileVisibility: String!
  showEmail: Boolean!
  showLocation: Boolean!
}

input CreateUserInput {
  username: String!
  email: String!
  password: String!
  role: UserRole = USER
  profile: CreateUserProfileInput
}

input CreateUserProfileInput {
  firstName: String
  lastName: String
  bio: String
  location: CreateLocationInput
}

input CreateLocationInput {
  country: String!
  city: String
  coordinates: CreateCoordinatesInput
}

input CreateCoordinatesInput {
  latitude: Float!
  longitude: Float!
}

input CreatePostInput {
  title: String!
  content: String!
  status: ContentStatus = DRAFT
  tags: [String!]
  media: [Upload!]
}

input UpdatePostInput {
  title: String
  content: String
  status: ContentStatus
  tags: [String!]
}

input CreateCommentInput {
  content: String!
  postId: ID!
  parentCommentId: ID
}

input NotificationFilter {
  priority: NotificationPriority
  read: Boolean
  type: String
  startDate: DateTime
  endDate: DateTime
}

type Query {
  node(id: ID!): Node
  user(id: ID!): User
  post(id: ID!): Post
  postsOld(filter: [ID!]) : [Post!]! @deprecated(reason: "Use posts instead")
  posts(filter: PostFilter): [Post!]!
  comments(postId: ID!): [Comment!]!
  notifications(filter: NotificationFilter): [Notification!]!
  search(query: String!): SearchResult!
}

type Mutation {
  createUser(input: CreateUserInput!): User!
  createPost(input: CreatePostInput!): Post!
  updatePost(id: ID!, input: UpdatePostInput!): Post!
  createComment(input: CreateCommentInput!): Comment!
  deletePost(id: ID!): Boolean!
  uploadMedia(file: Upload!): Media!
  updateUserPreferences(id: ID!, preferences: UserPreferencesInput!): UserPreferences!
}

type Subscription {
  postUpdated(id: ID!): Post!
  newComment(postId: ID!): Comment!
  notificationReceived(userId: ID!): Notification!
}

union SearchResult = User | Post | Comment | Tag

input PostFilter {
  status: ContentStatus
  authorId: ID
  tags: [String!]
  dateRange: DateRangeInput
}

input DateRangeInput {
  start: DateTime!
  end: DateTime!
}

input UserPreferencesInput {
  theme: String
  language: String
  notifications: NotificationPreferencesInput
  privacy: PrivacySettingsInput
}

input NotificationPreferencesInput {
  email: Boolean
  push: Boolean
  sms: Boolean
  frequency: String
}

input PrivacySettingsInput {
  profileVisibility: String
  showEmail: Boolean
  showLocation: Boolean
}

directive @auth(requires: UserRole!) on FIELD_DEFINITION
directive @cache(ttl: Int!) on FIELD_DEFINITION
directive @deprecated(reason: String) on FIELD_DEFINITION 