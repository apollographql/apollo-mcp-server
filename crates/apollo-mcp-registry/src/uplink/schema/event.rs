use crate::uplink::schema::SchemaState;
use std::fmt::Debug;
use std::fmt::Formatter;

/// Schema events
pub enum Event {
    /// The schema was updated.
    UpdateSchema(SchemaState),

    /// There are no more updates to the schema
    NoMoreSchema,
}

impl Debug for Event {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Event::UpdateSchema(_) => {
                write!(f, "UpdateSchema(<redacted>)")
            }
            Event::NoMoreSchema => {
                write!(f, "NoMoreSchema")
            }
        }
    }
}
