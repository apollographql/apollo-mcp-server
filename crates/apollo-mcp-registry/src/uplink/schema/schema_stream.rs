// tonic does not derive `Eq` for the gRPC message types, which causes a warning from Clippy. The
// current suggestion is to explicitly allow the lint in the module that imports the protos.
// Read more: https://github.com/hyperium/tonic/issues/1056
#![allow(clippy::derive_partial_eq_without_eq)]

use crate::uplink::UplinkRequest;
use crate::uplink::UplinkResponse;
use crate::uplink::schema::SchemaState;
use crate::uplink::schema::schema_stream::supergraph_sdl_query::FetchErrorCode;
use crate::uplink::schema::schema_stream::supergraph_sdl_query::SupergraphSdlQueryRouterConfig;
use graphql_client::GraphQLQuery;

#[derive(GraphQLQuery)]
#[graphql(
    query_path = "src/uplink/schema/schema_query.graphql",
    schema_path = "src/uplink/uplink.graphql",
    request_derives = "Debug",
    response_derives = "PartialEq, Debug, Deserialize",
    deprecated = "warn"
)]
pub(crate) struct SupergraphSdlQuery;

impl From<UplinkRequest> for supergraph_sdl_query::Variables {
    fn from(req: UplinkRequest) -> Self {
        supergraph_sdl_query::Variables {
            api_key: req.api_key,
            graph_ref: req.graph_ref,
            if_after_id: req.id,
        }
    }
}

impl From<supergraph_sdl_query::ResponseData> for UplinkResponse<String> {
    fn from(response: supergraph_sdl_query::ResponseData) -> Self {
        match response.router_config {
            SupergraphSdlQueryRouterConfig::RouterConfigResult(result) => UplinkResponse::New {
                response: result.supergraph_sdl,
                id: result.id,
                // this will truncate the number of seconds to under u64::MAX, which should be
                // a large enough delay anyway
                delay: result.min_delay_seconds as u64,
            },
            SupergraphSdlQueryRouterConfig::Unchanged(response) => UplinkResponse::Unchanged {
                id: Some(response.id),
                delay: Some(response.min_delay_seconds as u64),
            },
            SupergraphSdlQueryRouterConfig::FetchError(err) => UplinkResponse::Error {
                retry_later: err.code == FetchErrorCode::RETRY_LATER,
                code: match err.code {
                    FetchErrorCode::AUTHENTICATION_FAILED => "AUTHENTICATION_FAILED".to_string(),
                    FetchErrorCode::ACCESS_DENIED => "ACCESS_DENIED".to_string(),
                    FetchErrorCode::UNKNOWN_REF => "UNKNOWN_REF".to_string(),
                    FetchErrorCode::RETRY_LATER => "RETRY_LATER".to_string(),
                    FetchErrorCode::NOT_IMPLEMENTED_ON_THIS_INSTANCE => {
                        "NOT_IMPLEMENTED_ON_THIS_INSTANCE".to_string()
                    }
                    FetchErrorCode::Other(other) => other,
                },
                message: err.message,
            },
        }
    }
}

impl From<supergraph_sdl_query::ResponseData> for UplinkResponse<SchemaState> {
    fn from(response: supergraph_sdl_query::ResponseData) -> Self {
        match response.router_config {
            SupergraphSdlQueryRouterConfig::RouterConfigResult(result) => UplinkResponse::New {
                response: SchemaState {
                    sdl: result.supergraph_sdl,
                    launch_id: Some(result.id.clone()),
                },
                id: result.id,
                // this will truncate the number of seconds to under u64::MAX, which should be
                // a large enough delay anyway
                delay: result.min_delay_seconds as u64,
            },
            SupergraphSdlQueryRouterConfig::Unchanged(response) => UplinkResponse::Unchanged {
                id: Some(response.id),
                delay: Some(response.min_delay_seconds as u64),
            },
            SupergraphSdlQueryRouterConfig::FetchError(err) => UplinkResponse::Error {
                retry_later: err.code == FetchErrorCode::RETRY_LATER,
                code: match err.code {
                    FetchErrorCode::AUTHENTICATION_FAILED => "AUTHENTICATION_FAILED".to_string(),
                    FetchErrorCode::ACCESS_DENIED => "ACCESS_DENIED".to_string(),
                    FetchErrorCode::UNKNOWN_REF => "UNKNOWN_REF".to_string(),
                    FetchErrorCode::RETRY_LATER => "RETRY_LATER".to_string(),
                    FetchErrorCode::NOT_IMPLEMENTED_ON_THIS_INSTANCE => {
                        "NOT_IMPLEMENTED_ON_THIS_INSTANCE".to_string()
                    }
                    FetchErrorCode::Other(other) => other,
                },
                message: err.message,
            },
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_uplink_request_to_graphql_variables() {
        let request = UplinkRequest {
            api_key: "test_key".to_string(),
            graph_ref: "test_ref".to_string(),
            id: Some("test_id".to_string()),
        };

        let variables: supergraph_sdl_query::Variables = request.into();

        assert_eq!(variables.api_key, "test_key");
        assert_eq!(variables.graph_ref, "test_ref");
        assert_eq!(variables.if_after_id, Some("test_id".to_string()));
    }

    #[test]
    fn test_graphql_response_to_uplink_response_new() {
        let response = supergraph_sdl_query::ResponseData {
            router_config: SupergraphSdlQueryRouterConfig::RouterConfigResult(
                supergraph_sdl_query::SupergraphSdlQueryRouterConfigOnRouterConfigResult {
                    supergraph_sdl: "test_sdl".to_string(),
                    id: "result_id".to_string(),
                    min_delay_seconds: 42.0,
                },
            ),
        };

        let uplink_response: UplinkResponse<String> = response.into();

        assert!(matches!(
            uplink_response,
            UplinkResponse::New { response, id, delay }
            if response == "test_sdl" && id == "result_id" && delay == 42
        ));
    }

    #[test]
    fn test_graphql_response_to_uplink_response_unchanged() {
        let response = supergraph_sdl_query::ResponseData {
            router_config: SupergraphSdlQueryRouterConfig::Unchanged(
                supergraph_sdl_query::SupergraphSdlQueryRouterConfigOnUnchanged {
                    id: "unchanged_id".to_string(),
                    min_delay_seconds: 30.0,
                },
            ),
        };

        let uplink_response: UplinkResponse<String> = response.into();

        assert!(matches!(
            uplink_response,
            UplinkResponse::Unchanged { id, delay }
            if id == Some("unchanged_id".to_string()) && delay == Some(30)
        ));
    }

    #[test]
    fn test_graphql_response_to_uplink_response_error() {
        let response = supergraph_sdl_query::ResponseData {
            router_config: SupergraphSdlQueryRouterConfig::FetchError(
                supergraph_sdl_query::SupergraphSdlQueryRouterConfigOnFetchError {
                    code: FetchErrorCode::RETRY_LATER,
                    message: "Try again later".to_string(),
                },
            ),
        };

        let uplink_response: UplinkResponse<String> = response.into();

        assert!(matches!(
            uplink_response,
            UplinkResponse::Error { retry_later, code, message }
            if retry_later && code == "RETRY_LATER" && message == "Try again later"
        ));
    }
}
