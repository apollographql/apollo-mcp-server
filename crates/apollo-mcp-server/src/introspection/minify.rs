use apollo_compiler::schema::{ExtendedType, Type};
use regex::Regex;
use std::{collections::HashMap, sync::OnceLock};

pub trait MinifyExt {
    /// Serialize in minified form
    fn minify(&self) -> String;
}

impl MinifyExt for ExtendedType {
    fn minify(&self) -> String {
        match self {
            ExtendedType::Scalar(scalar_type) => minify_scalar(scalar_type),
            ExtendedType::Object(object_type) => minify_object(object_type),
            ExtendedType::Interface(interface_type) => minify_interface(interface_type),
            ExtendedType::Union(union_type) => minify_union(union_type),
            ExtendedType::Enum(enum_type) => minify_enum(enum_type),
            ExtendedType::InputObject(input_object_type) => minify_input_object(input_object_type),
        }
    }
}

fn minify_scalar(scalar_type: &apollo_compiler::schema::ScalarType) -> String {
    shorten_scalar_names(scalar_type.name.as_str()).to_string()
}

fn minify_object(object_type: &apollo_compiler::schema::ObjectType) -> String {
    let fields = minify_fields(&object_type.fields);
    let type_name = format_type_name_with_description(&object_type.name, &object_type.description);
    let interfaces = format_interfaces(&object_type.implements_interfaces);

    if interfaces.is_empty() {
        format!("T:{type_name}:{fields}")
    } else {
        format!("T:{type_name}<{interfaces}>:{fields}")
    }
}

fn minify_interface(interface_type: &apollo_compiler::schema::InterfaceType) -> String {
    let fields = minify_fields(&interface_type.fields);
    let type_name =
        format_type_name_with_description(&interface_type.name, &interface_type.description);
    format!("F:{type_name}:{fields}")
}

fn minify_union(union_type: &apollo_compiler::schema::UnionType) -> String {
    let member_types = union_type
        .members
        .iter()
        .map(|member| member.as_str())
        .collect::<Vec<&str>>()
        .join(",");
    let type_name = format_type_name_with_description(&union_type.name, &union_type.description);
    format!("U:{type_name}:{member_types}")
}

fn minify_enum(enum_type: &apollo_compiler::schema::EnumType) -> String {
    let values = enum_type
        .values
        .keys()
        .map(|value| value.as_str())
        .collect::<Vec<&str>>()
        .join(",");
    let type_name = format_type_name_with_description(&enum_type.name, &enum_type.description);
    format!("E:{type_name}:{values}")
}

fn minify_input_object(input_object_type: &apollo_compiler::schema::InputObjectType) -> String {
    let fields = minify_input_fields(&input_object_type.fields);
    let type_name =
        format_type_name_with_description(&input_object_type.name, &input_object_type.description);
    format!("I:{type_name}:{fields}")
}

// We should only minify directives that assist the LLM in understanding the schema. This included @deprecated
fn minify_directives(directives: &apollo_compiler::ast::DirectiveList) -> String {
    let mut result = String::new();

    static DIRECTIVES_TO_MINIFY: OnceLock<HashMap<&str, &str>> = OnceLock::new();
    let directives_to_minify =
        DIRECTIVES_TO_MINIFY.get_or_init(|| HashMap::from([("deprecated", "D")]));

    for directive in directives.iter() {
        if let Some(minified_name) = directives_to_minify.get(directive.name.as_str()) {
            if !directive.arguments.is_empty() {
                // Since we're only handling @deprecated right now we can just add the reason and minify it
                let reason = directive
                    .arguments
                    .iter()
                    .find(|a| a.name == "reason")
                    .and_then(|a| a.value.as_str())
                    .unwrap_or("No longer supported")
                    .to_string();
                result.push_str(&format!("@{}(\"{}\")", minified_name, reason));
            } else {
                result.push_str(&format!("@{}", minified_name));
            }
        }
    }
    result
}

fn minify_fields(
    fields: &apollo_compiler::collections::IndexMap<
        apollo_compiler::Name,
        apollo_compiler::schema::Component<apollo_compiler::ast::FieldDefinition>,
    >,
) -> String {
    let mut result = String::new();

    for (field_name, field) in fields.iter() {
        // Add description if present
        if let Some(desc) = field.description.as_ref() {
            result.push_str(&format!("\"{}\"", normalize_description(desc)));
        }

        // Add field name
        result.push_str(field_name.as_str());

        // Add arguments if present
        if !field.arguments.is_empty() {
            result.push('(');
            result.push_str(&minify_arguments(&field.arguments));
            result.push(')');
        }

        // Add field type
        result.push(':');
        result.push_str(&type_name(&field.ty));
        result.push_str(&minify_directives(&field.directives));

        result.push(',');
    }

    // Remove trailing comma
    if !result.is_empty() {
        result.pop();
    }

    result
}

fn minify_input_fields(
    fields: &apollo_compiler::collections::IndexMap<
        apollo_compiler::Name,
        apollo_compiler::schema::Component<apollo_compiler::ast::InputValueDefinition>,
    >,
) -> String {
    let mut result = String::new();

    for (field_name, field) in fields.iter() {
        // Add description if present
        if let Some(desc) = field.description.as_ref() {
            result.push_str(&format!("\"{}\"", normalize_description(desc)));
        }

        // Add field name and type
        result.push_str(field_name.as_str());
        result.push(':');
        result.push_str(&type_name(&field.ty));
        result.push_str(&minify_directives(&field.directives));
        result.push(',');
    }

    // Remove trailing comma
    if !result.is_empty() {
        result.pop();
    }

    result
}

fn minify_arguments(
    arguments: &[apollo_compiler::Node<apollo_compiler::ast::InputValueDefinition>],
) -> String {
    arguments
        .iter()
        .map(|arg| {
            if let Some(desc) = arg.description.as_ref() {
                format!(
                    "\"{}\"{}:{}{}",
                    normalize_description(desc),
                    arg.name.as_str(),
                    type_name(&arg.ty),
                    minify_directives(&arg.directives)
                )
            } else {
                format!(
                    "{}:{}{}",
                    arg.name.as_str(),
                    type_name(&arg.ty),
                    minify_directives(&arg.directives)
                )
            }
        })
        .collect::<Vec<String>>()
        .join(",")
}

fn format_type_name_with_description(
    name: &apollo_compiler::Name,
    description: &Option<apollo_compiler::Node<str>>,
) -> String {
    if let Some(desc) = description.as_ref() {
        format!("\"{}\"{}", normalize_description(desc), name)
    } else {
        name.to_string()
    }
}

fn format_interfaces(
    interfaces: &apollo_compiler::collections::IndexSet<apollo_compiler::schema::ComponentName>,
) -> String {
    interfaces
        .iter()
        .map(|interface| interface.as_str())
        .collect::<Vec<&str>>()
        .join(",")
}

fn type_name(ty: &Type) -> String {
    let name = shorten_scalar_names(ty.inner_named_type().as_str());
    if ty.is_list() {
        format!("[{name}]")
    } else if ty.is_non_null() {
        format!("{name}!")
    } else {
        name.to_string()
    }
}

fn shorten_scalar_names(name: &str) -> &str {
    match name {
        "String" => "s",
        "Int" => "i",
        "Float" => "f",
        "Boolean" => "b",
        "ID" => "d",
        _ => name,
    }
}

/// Normalize description formatting
#[allow(clippy::expect_used)]
fn normalize_description(desc: &str) -> String {
    // LLMs can typically process descriptions just fine without whitespace
    static WHITESPACE_PATTERN: OnceLock<Regex> = OnceLock::new();
    let re = WHITESPACE_PATTERN.get_or_init(|| Regex::new(r"\s+").expect("regex pattern compiles"));
    re.replace_all(desc, "").to_string()
}
