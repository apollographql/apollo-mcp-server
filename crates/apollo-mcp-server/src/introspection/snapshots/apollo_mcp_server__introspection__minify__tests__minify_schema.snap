---
source: crates/apollo-mcp-server/src/introspection/minify.rs
expression: minified
---
__Schema: T:"AGraphQLSchemadefinesthecapabilitiesofaGraphQLserver.Itexposesallavailabletypesanddirectivesontheserver,aswellastheentrypointsforquery,mutation,andsubscriptionoperations."__Schema:description:s,"Alistofalltypessupportedbythisserver."types:[__Type],"Thetypethatqueryoperationswillberootedat."queryType:__Type!,"Ifthisserversupportsmutation,thetypethatmutationoperationswillberootedat."mutationType:__Type,"Ifthisserversupportsubscription,thetypethatsubscriptionoperationswillberootedat."subscriptionType:__Type,"Alistofalldirectivessupportedbythisserver."directives:[__Directive]
__Type: T:"ThefundamentalunitofanyGraphQLSchemaisthetype.TherearemanykindsoftypesinGraphQLasrepresentedbythe`__TypeKind`enum.Dependingonthekindofatype,certainfieldsdescribeinformationaboutthattype.Scalartypesprovidenoinformationbeyondaname,descriptionandoptional`specifiedByURL`,whileEnumtypesprovidetheirvalues.ObjectandInterfacetypesprovidethefieldstheydescribe.Abstracttypes,UnionandInterface,providetheObjecttypespossibleatruntime.ListandNonNulltypescomposeothertypes."__Type:kind:__TypeKind!,name:s,description:s,fields(includeDeprecated:b):[__Field],interfaces:[__Type],possibleTypes:[__Type],enumValues(includeDeprecated:b):[__EnumValue],inputFields(includeDeprecated:b):[__InputValue],ofType:__Type,specifiedByURL:s
__TypeKind: E:"Anenumdescribingwhatkindoftypeagiven`__Type`is."__TypeKind:SCALAR,OBJECT,INTERFACE,UNION,ENUM,INPUT_OBJECT,LIST,NON_NULL
__Field: T:"ObjectandInterfacetypesaredescribedbyalistofFields,eachofwhichhasaname,potentiallyalistofarguments,andareturntype."__Field:name:s!,description:s,args(includeDeprecated:b):[__InputValue],type:__Type!,isDeprecated:b!,deprecationReason:s
__InputValue: T:"ArgumentsprovidedtoFieldsorDirectivesandtheinputfieldsofanInputObjectarerepresentedasInputValueswhichdescribetheirtypeandoptionallyadefaultvalue."__InputValue:name:s!,description:s,type:__Type!,"AGraphQL-formattedstringrepresentingthedefaultvalueforthisinputvalue."defaultValue:s,isDeprecated:b!,deprecationReason:s
__EnumValue: T:"OnepossiblevalueforagivenEnum.Enumvaluesareuniquevalues,notaplaceholderforastringornumericvalue.HoweveranEnumvalueisreturnedinaJSONresponseasastring."__EnumValue:name:s!,description:s,isDeprecated:b!,deprecationReason:s
__Directive: T:"ADirectiveprovidesawaytodescribealternateruntimeexecutionandtypevalidationbehaviorinaGraphQLdocument.Insomecases,youneedtoprovideoptionstoalterGraphQL'sexecutionbehaviorinwaysfieldargumentswillnotsuffice,suchasconditionallyincludingorskippingafield.Directivesprovidethisbydescribingadditionalinformationtotheexecutor."__Directive:name:s!,description:s,locations:[__DirectiveLocation],args(includeDeprecated:b):[__InputValue],isRepeatable:b!
__DirectiveLocation: E:"ADirectivecanbeadjacenttomanypartsoftheGraphQLlanguage,a__DirectiveLocationdescribesonesuchpossibleadjacencies."__DirectiveLocation:QUERY,MUTATION,SUBSCRIPTION,FIELD,FRAGMENT_DEFINITION,FRAGMENT_SPREAD,INLINE_FRAGMENT,VARIABLE_DEFINITION,SCHEMA,SCALAR,OBJECT,FIELD_DEFINITION,ARGUMENT_DEFINITION,INTERFACE,UNION,ENUM,ENUM_VALUE,INPUT_OBJECT,INPUT_FIELD_DEFINITION
Int: i
Float: f
String: s
Boolean: b
ID: d
DateTime: DateTime
JSON: JSON
Upload: Upload
UserRole: E:UserRole:ADMIN,MODERATOR,USER,GUEST
ContentStatus: E:ContentStatus:DRAFT,PUBLISHED,ARCHIVED,DELETED
NotificationPriority: E:NotificationPriority:LOW,MEDIUM,HIGH,URGENT
MediaType: E:MediaType:IMAGE,VIDEO,AUDIO,DOCUMENT
Node: F:Node:id:d!,createdAt:DateTime!,updatedAt:DateTime!
Content: F:Content:id:d!,title:s!,status:ContentStatus!,author:User!,metadata:JSON
User: T:User<Node>:id:d!,createdAt:DateTime!,updatedAt:DateTime!,username:s!,email:s!,role:UserRole!,profile:UserProfile,posts:[Post],comments:[Comment],notifications:[Notification],preferences:UserPreferences!
UserProfile: T:UserProfile:firstName:s,lastName:s,bio:s,avatar:Media,socialLinks:[SocialLink],location:Location
Location: T:Location:country:s!,city:s,coordinates:Coordinates
Coordinates: T:Coordinates:latitude:f!,longitude:f!
SocialLink: T:SocialLink:platform:s!,url:s!,verified:b!
Post: T:Post<Node,Content>:id:d!,createdAt:DateTime!,updatedAt:DateTime!,title:s!,content:s!,status:ContentStatus!,author:User!,metadata:JSON,comments:[Comment],media:[Media],tags:[Tag],analytics:PostAnalytics!
Comment: T:Comment<Node>:id:d!,createdAt:DateTime!,updatedAt:DateTime!,content:s!,author:User!,post:Post!,parentComment:Comment,replies:[Comment],reactions:[Reaction]
Media: T:Media:id:d!,type:MediaType!,url:s!,thumbnail:s,metadata:MediaMetadata!,uploader:User!
MediaMetadata: T:MediaMetadata:size:i!,format:s!,dimensions:Dimensions,duration:i
Dimensions: T:Dimensions:width:i!,height:i!
Tag: T:Tag:id:d!,name:s!,slug:s!,description:s,posts:[Post]
Reaction: T:Reaction:id:d!,type:s!,user:User!,comment:Comment!,createdAt:DateTime!
Notification: T:Notification:id:d!,type:s!,priority:NotificationPriority!,message:s!,recipient:User!,read:b!,createdAt:DateTime!,metadata:JSON
PostAnalytics: T:PostAnalytics:views:i!,likes:i!,shares:i!,comments:i!,engagement:f!,demographics:Demographics!
Demographics: T:Demographics:ageGroups:[AgeGroup],locations:[LocationStats],devices:[DeviceStats]
AgeGroup: T:AgeGroup:range:s!,percentage:f!
LocationStats: T:LocationStats:country:s!,count:i!
DeviceStats: T:DeviceStats:type:s!,count:i!
UserPreferences: T:UserPreferences:theme:s!,language:s!,notifications:NotificationPreferences!,privacy:PrivacySettings!
NotificationPreferences: T:NotificationPreferences:email:b!,push:b!,sms:b!,frequency:s!
PrivacySettings: T:PrivacySettings:profileVisibility:s!,showEmail:b!,showLocation:b!
CreateUserInput: I:CreateUserInput:username:s!,email:s!,password:s!,role:UserRole,profile:CreateUserProfileInput
CreateUserProfileInput: I:CreateUserProfileInput:firstName:s,lastName:s,bio:s,location:CreateLocationInput
CreateLocationInput: I:CreateLocationInput:country:s!,city:s,coordinates:CreateCoordinatesInput
CreateCoordinatesInput: I:CreateCoordinatesInput:latitude:f!,longitude:f!
CreatePostInput: I:CreatePostInput:title:s!,content:s!,status:ContentStatus,tags:[s],media:[Upload]
UpdatePostInput: I:UpdatePostInput:title:s,content:s,status:ContentStatus,tags:[s]
CreateCommentInput: I:CreateCommentInput:content:s!,postId:d!,parentCommentId:d
NotificationFilter: I:NotificationFilter:priority:NotificationPriority,read:b,type:s,startDate:DateTime,endDate:DateTime
Query: T:Query:node(id:d!):Node,user(id:d!):User,post(id:d!):Post,postsOld(filter:[d]):[Post]@D("Usepostsinstead"),posts(filter:PostFilter):[Post],comments(postId:d!):[Comment],notifications(filter:NotificationFilter):[Notification],search(query:s!):SearchResult!
Mutation: T:Mutation:createUser(input:CreateUserInput!):User!,createPost(input:CreatePostInput!):Post!,updatePost(id:d!,input:UpdatePostInput!):Post!,createComment(input:CreateCommentInput!):Comment!,deletePost(id:d!):b!,uploadMedia(file:Upload!):Media!,updateUserPreferences(id:d!,preferences:UserPreferencesInput!):UserPreferences!
Subscription: T:Subscription:postUpdated(id:d!):Post!,newComment(postId:d!):Comment!,notificationReceived(userId:d!):Notification!
SearchResult: U:SearchResult:User,Post,Comment,Tag
PostFilter: I:PostFilter:status:ContentStatus,authorId:d,tags:[s],dateRange:DateRangeInput
DateRangeInput: I:DateRangeInput:start:DateTime!,end:DateTime!
UserPreferencesInput: I:UserPreferencesInput:theme:s,language:s,notifications:NotificationPreferencesInput,privacy:PrivacySettingsInput
NotificationPreferencesInput: I:NotificationPreferencesInput:email:b,push:b,sms:b,frequency:s
PrivacySettingsInput: I:PrivacySettingsInput:profileVisibility:s,showEmail:b,showLocation:b
